src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
+            if !cpu.mems.is_empty() {
+                // res.cpu.mems = cpu.mems.clone();
+                cpuset_controller.set_mems(&cpu.mems);
+            }
 
-    Ok(())
-}
+            let cpu_controller: &CpuController = cg.controller_of().unwrap();
 
-fn read_file(dir: &str, file: &str) -> Result<String> {
-    let p = format!("{}/{}", dir, file);
-    let ret = fs::read_to_string(p.as_str())?;
-    Ok(ret)
-}
+            let mut shares = cpu.shares.unwrap_or(0);
+            if shares != 0 {
+                if cg.v2() {
+                    shares = convert_shares_to_v2_value(shares);
+                }
+                cpu_controller.set_shares(shares);
+            }
 
-fn copy_parent(dir: &str, file: &str) -> Result<()> {
-    let parent = if let Some(index) = dir.rfind('/') {
-        &dir[..index]
-    } else {
-        return Err(anyhow!("cannot copy file from parent"));
-    };
+            let quota = cpu.quota.unwrap_or(0) as u64;
+            let period = cpu.period.unwrap_or(0);
+            cpu_controller.set_cfs_quota_and_period(quota, period);
 
-    match read_file(parent, file) {
-        Ok(v) => {
-            if !v.trim().is_empty() {
-                info!(sl!(), "value: \"{}\"", v.as_str().trim());
-                return write_file(dir, file, v);
-            } else {
-                copy_parent(parent, file)?;
-                return copy_parent(dir, file);
+            let realtime_runtime = cpu.realtime_runtime.unwrap_or(0);
+            if realtime_runtime != 0 {
+                cpu_controller.set_rt_runtime(realtime_runtime);
+            }
+            let realtime_period = cpu.realtime_period.unwrap_or(0);
+            if realtime_period != 0 {
+                cpu_controller.set_rt_period_us(realtime_period);
             }
         }
-        Err(ref e) if io_error_kind_eq(e, std::io::ErrorKind::NotFound) => {
-            copy_parent(parent, file)?;
-            return copy_parent(dir, file);
-        }
-        Err(e) => return Err(e.into()),
-    }
-}
 
-fn write_nonzero(dir: &str, file: &str, v: i128) -> Result<()> {
-    if v != 0 {
-        write_file(dir, file, v.to_string())?;
-    }
+        if r.memory.is_some() {
+            info!(sl!(), "cgroup manager set memory !!!");
+            let mem_controller: &MemController = cg.controller_of().unwrap();
+            let memory = r.memory.as_ref().unwrap();
 
-    Ok(())
-}
+            if !update {
+                // initialize kmem limits for accounting
+                mem_controller.set_kmem_limit(1);
+                mem_controller.set_kmem_limit(-1);
+            }
 
-fn try_write_nonzero(dir: &str, file: &str, v: i128) -> Result<()> {
-    match write_nonzero(dir, file, v) {
-        Ok(_) => Ok(()),
-        Err(ref e) if io_error_kind_eq(e, std::io::ErrorKind::PermissionDenied) => Ok(()),
-        Err(e) => Err(anyhow!(e)),
-    }
-}
+            let limit = memory.limit.unwrap_or(0);
```

Comment:
In my opinion, many of the cases below are quite close, if not almost identical.

* they take some value
* they do an `_unwrap_or(defaultValue)`
* they apply using some associated controller
* that controller may fail, in which case I think we should report (in most cases we don't in the current version)

This is so similar in all cases that it sort of calls for a macro to do the grunt work, to reduce copy-paste inheritance.

Reply:
Yes, I will create a macro to reduce some codes.